src/PMGDQueryHandler.cc

```diff
@@ -53,17 +54,24 @@ void PMGDQueryHandler::init()
     _db = new PMGD::Graph(dbname.c_str(), PMGD::Graph::Create);
 
     // Create the query handler here assuming database is valid now.
-    _dblock = new std::mutex();
+    _dblock = new RWLock();
 }
 
 std::vector<PMGDCmdResponses>
               PMGDQueryHandler::process_queries(const PMGDCmds &cmds,
-              int num_groups)
+              int num_groups, bool readonly)
 {
     std::vector<PMGDCmdResponses> responses(num_groups);
 
     assert(_tx == NULL);
-    _dblock->lock();
+
+    // Assuming one query handler handles one TX at a time.
```

Comment:
Should there be a check for this?

Reply:
We are safe with this one. It will require a lot more design changes to affect that.